geopandas/clip.py

```diff
@@ -0,0 +1,229 @@
+"""
+geopandas.clip
+============
+
+A module to clip vector data using GeoPandas.
+
+"""
+
+import pandas as pd
+import geopandas as gpd
+
+
+def _clip_points(shp, clip_obj):
+    """Clip point geometry to the clip_obj GeoDataFrame extent.
+
+    Clip an input point GeoDataFrame to the polygon extent of the clip_obj
+    parameter. Points that intersect the clip_obj geometry are extracted with
+    associated attributes and returned.
+
+    Parameters
+    ----------
+    shp : GeoDataFrame
+        Composed of point geometry that is clipped to clip_obj.
+
+    clip_obj : GeoDataFrame
+        Reference polygon for clipping.
+
+    Returns
+    -------
+    GeoDataFrame
+        The returned GeoDataFrame is a subset of shp that intersects
+        with clip_obj.
+    """
+    poly = clip_obj.geometry.unary_union
```

Comment:
i haven't! i think it would be interesting to do more extensive performance testing and would love to understand how @jorisvandenbossche and everyone implements that. we found clipping points to be fairly efficient BUT i know this could quickly fail when you get into the hundreds of thousands to millions of points. i'd personally love to see this somehow be parallelized if possible but that is beyond my capacity at this point i think to contribute. (just with time). open to suggestions.

Reply:
The same code you are using below should do the trick. I haven't done any performance tests, but for large number of points, especially if large number of them is outside clipping polygon, this might be faster.

```suggestion
    poly = clip_obj.geometry.unary_union
    
    spatial_index = gdf.sindex
    bbox = poly.bounds
    sidx = list(spatial_index.intersection(bbox))
    gdf_sub = gdf.iloc[sidx]
    
    return gdf_sub[gdf_sub.geometry.intersects(poly)]
```

Reply:
you are probably right!! i'll commit this and will test quickly locally. thank you @martinfleis  !!!